File: 2021/day_2/day_2.R
library(tidyverse)
library(slider)

d <- read_table("2021/day_2/input.txt", col_names = c("dir", "len"))


# part 1 ------------------------------------------------------------------

d %>% 
  group_by(dir) %>% 
  summarise(total_len = sum(len)) %>% 
  pivot_wider(names_from = dir, values_from = total_len) %>% 
  mutate(ans = (down - up)*forward)


# part 2 loop ------------------------------------------------------------------

aim <- 0
horiz <- 0
depth <- 0

for (i in seq_along(d$dir)) {
  
  if (d$dir[i] == "up"){
    aim <- aim - d$len[i]
  }
  
  if (d$dir[i] == "down"){
    aim <- aim + d$len[i]
  }
  
  if (d$dir[i] == "forward"){
    horiz <- horiz + d$len[i]
    depth <- depth + d$len[i] * aim
  }
  
}

horiz*depth

# part 2 tidy -------------------------------------------------------------

d %>% 
  mutate(len = if_else(dir == "up", -len, len),
         aim = if_else(dir == "forward", 0, len),
         cumu_aim = slide_sum(aim, before = Inf)) %>% 
  filter(dir == "forward") %>% 
  summarise(depth = sum(len*cumu_aim),
            horiz = sum(len),
            ans = depth*horiz)

File: 2021/day_2/day_2.py
import pandas as pd

dd = pd.read_table("2021/day_2/input.txt", sep = " ", names=["dir", "len"])
# Part 1

dd["len"] = pd.to_numeric(dd["len"])

final_depth = sum(dd["len"][dd["dir"] == "down"]) - sum(dd["len"][dd["dir"] == "up"])

final_horiz = sum(dd["len"][dd["dir"] == "forward"]) 

ans1 = final_depth * final_horiz

# Part 2
aim = 0
horiz = 0
depth = 0

dd["dir"][1] == "down"

for i in range(len(dd["dir"])):
    if dd["dir"][i] == "down":
        aim += dd["len"][i]
    else:
        if dd["dir"][i] == "up":
            aim += -dd["len"][i]
        else:
            if dd["dir"][i] == "forward":
                horiz += dd["len"][i] 
                depth += dd["len"][i] * aim


ans2 = horiz*depth

print(f'The answer to Part 1 is {ans1}.\nThe answer to Part 2 is {ans2}.')